torch/distributed/fsdp/_runtime_utils.py

```diff
@@ -512,35 +504,35 @@ def _post_forward(
     parameter.
     """
     with torch.profiler.record_function("FullyShardedDataParallel._post_forward"):
-        state._exec_order_data.record_post_forward(handles)
+        state._exec_order_data.record_post_forward(handle)
         if reshard_fn is not None:
-            reshard_fn()
+            reshard_fn(state, handle)
         # Register pre-backward hooks to unshard the flat parameters for the
         # gradient computation (if needed)
-        output = _register_pre_backward_hooks(state, module, output, handles)
+        output = _register_pre_backward_hooks(state, module, output, handle)
         state.training_state = TrainingState.IDLE
-        for handle in handles:
+        if handle:
             handle._training_state = HandleTrainingState.IDLE
         return output
 
 
 @no_type_check
 def _post_forward_reshard(
     state: _FSDPState,
-    handles: List[FlatParamHandle],
+    handle: FlatParamHandle,
 ) -> None:
     """Reshards parameters in the post-forward."""
-    if not handles:
+    if not handle:
         return
     # Do not free the root's parameters in the post-forward for `FULL_SHARD`
     # with the intention that they are immediately used for backward
     # computation (though this may not be true)
-    free_unsharded_flat_params = [
-        not state._is_root
+    free_unsharded_flat_param = (
+        handle
```

Comment:
nit: do not need the `handle` in this condition since we return early on `if not handle:`